### ISSUE TITLE
Request Deduplication

### ISSUE THREAD

**Issue by h3rmanj** [AzureAD/microsoft-authentication-library-for-js#4686]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Wrapper Library

Not Applicable

### Description

Hi!

I have a React project which uses `@azure/msal-browser`. When we call our API, we use a custom fetcher that calls `acquireTokenSilent` before the request. We use the custom fetcher through a library (`SWR`) that includes fetching hooks with request deduplication etc.

So if I use the fetching hook from the library multiple places to fetch data from the same endpoint, our custom fetcher will only be called once. However, if the hook is used multiple times in the same page, to different endpoints, our custom fetcher gets called multiple times, and therefore also `acquireTokenSilent`. This results in multiple network calls for configurations and tokens, and often so many that the browser limits other requests.

(I know i can skip the metadata retrieval by manually providing it, but it still happens for tokens)

![image](https://user-images.githubusercontent.com/11387548/162155552-c02bff88-373d-46d3-8956-d0b28d8fda41.png)
![image](https://user-images.githubusercontent.com/11387548/162155631-d901db96-91b3-4388-82db-e47eb21b807f.png)
![image](https://user-images.githubusercontent.com/11387548/162155978-cb6b32b1-3449-49fc-847b-cef0c25c2623.png)

I might be implementing the whole thing wrong, and should decouple the token request from the api request in my code. But it would be nice if there was some kind of request deduplication functionality in the package.

### Source

External (Customer)

**Comment by ellymakuba** [comment]
For instance discovery and openid-confiuration calls we are working on enhancements to reduce the number of network calls made by the library. For the token endpoint can you please confirm that you are requesting multiple different access tokens for different resources? Please share relevant code snippets on how you are invoking acquiretokensilent along with verbose logs from msal.

**Comment by tnorling** [comment]
The metadata call should only be made once per page load per `PublicClientApplication` instance. Can you make sure you are not reinstantiating `PublicClientApplication`? When using React you should not instantiate `PublicClientApplication` inside a React component due to re-renders potentially causing reinstantiation.

**Comment by h3rmanj** [comment]
Hi! Thanks for the quick answers. I used a self-made msal -> react binding package that probably was a little outdated. I did a long overdue refactor to use `@azure/msal-react`, and there's now less requests to retrieve the metadata. But there's still one per `acquireTokenSilent` request per resource.

I made a codesandbox to demonstrate this: https://codesandbox.io/s/gallant-liskov-ofrnlv?file=/src/App.js

It does two requests, one to retrieve a graph user, and one to retrieve subscriptions from azure. This can be changed if you want, please fork it and add your own client ids that does request to other stuff (important thing is it has to be different resources).

To reproduce, go to the code sandbox live preview site (https://ofrnlv.csb.app/), open the network tab, and login.

It will now do one request to the `instance?` and `openid-configuration`, and two separate request to retrieve access tokens, just as expected.

However, if you refresh, you'll now see multiple request to the `instance` and `openid-configuration` metadata endpoints. Thats the requests that should be deduplicated.

**Comment by jasonnutter** [comment]
@h3rmanj Thanks. What you observe is currently expected behavior. When you login, the login call will initiate the first call to teh metadata endpoints, which gets cached and reused for the `acquireTokenSilent` calls. When you are already logged in (after refreshing), the login call doesn't happen again, and since the two `acquireTokenSilent` calls happen in parallel, they'll both make a metadata network call. 

As mentioned, we have started working on improvements to how we fetch metadata, which I think will address the concerns here (i.e. by not making the network calls at all in most scenarios). You can follow #4536 to see our progress there.

**Comment by jasonnutter** [comment]
This should be addressed in `2.27.0`. Please open a new issue if there are further concerns.
